backends/aiger2/aiger.cc

```diff
@@ -0,0 +1,785 @@
+/*
+ *  yosys -- Yosys Open SYnthesis Suite
+ *
+ *  Copyright (C)  Martin Povišer <povik@cutebit.org>
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+// TODOs:
+//  - gracefully handling inout ports (an error message probably)
+
+#include "kernel/register.h"
+#include "kernel/celltypes.h"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+#define BITWISE_OPS ID($buf), ID($not), ID($mux), ID($and), ID($or), ID($xor), ID($xnor), ID($fa), \
+					ID($bwmux)
+
+#define REDUCE_OPS ID($reduce_and), ID($reduce_or), ID($reduce_xor), ID($reduce_xnor), ID($reduce_bool)
+
+#define LOGIC_OPS ID($logic_and), ID($logic_or), ID($logic_not)
+
+#define GATE_OPS ID($_BUF_), ID($_NOT_), ID($_AND_), ID($_NAND_), ID($_OR_), ID($_NOR_), \
+				 ID($_XOR_), ID($_XNOR_), ID($_ANDNOT_), ID($_ORNOT_), ID($_MUX_), ID($_NMUX_), \
+				 ID($_AOI3_), ID($_OAI3_), ID($_AOI4_), ID($_OAI4_)
+
+#define CMP_OPS ID($eq), ID($ne), ID($lt), ID($le), ID($ge), ID($gt)
+
+// TODO
+//#define ARITH_OPS ID($add), ID($sub), ID($neg)
+
+#define KNOWN_OPS BITWISE_OPS, REDUCE_OPS, LOGIC_OPS, GATE_OPS, ID($pos), CMP_OPS /*, ARITH_OPS*/
+
+template<typename Writer, typename Lit>
+struct Index {
+	static constexpr Lit CFALSE = Writer::CONST_FALSE;
+	static constexpr Lit CTRUE = Writer::CONST_TRUE;
```

Comment:
From the VS build:
```
 "D:\a\yosys\yosys\yosys-win32-vcxsrc-latest\YosysVS.sln" (default target) (1) ->
"D:\a\yosys\yosys\yosys-win32-vcxsrc-latest\YosysVS\YosysVS.vcxproj" (default target) (2) ->
(ClCompile target) -> 
  D:\a\yosys\yosys\yosys-win32-vcxsrc-latest\yosys\backends\aiger2\aiger.cc(49,40): error C2039: 'CONST_FALSE': is not a member of '`anonymous-namespace'::AigerWriter' [D:\a\yosys\yosys\yosys-win32-vcxsrc-latest\YosysVS\YosysVS.vcxproj]
  D:\a\yosys\yosys\yosys-win32-vcxsrc-latest\yosys\backends\aiger2\aiger.cc(49,1): error C2065: 'CONST_FALSE': undeclared identifier [D:\a\yosys\yosys\yosys-win32-vcxsrc-latest\YosysVS\YosysVS.vcxproj]
  ```
Not sure what to think of it.

Reply:
Doesn't it boil down to [this?](https://godbolt.org/z/qTb1bdoWP)

Reply:
Yes, thanks. It's not clear to me if it's valid C++, or how to change it to make the compiler happy

Reply:
I was told "default member initializers are complete-class context, so I think msvc is wrong to reject this" by a clang dev. Seems to suggest that initializers aren't declared until the class is complete, which it isn't because you have something like a type dependency loop. I guess you have to find a different class structure for the functionality you want. It's not too surprising to me that there would be problems related to `struct A : B<A>`